### PR TITLE
Code style and vs hints

### PR DESCRIPTION
**Pull request by shargon**

### CODE REVIEW

**Comment by shargon** [comment]
@erikzhang `StackItem.Null` is not used, can i remove it?

**Comment by igormcoelho** [comment]
@shargon let's try to port code conventions here before core project? neo-vm is shorter, we can do it quickly.

**Comment by erikzhang** [comment]
We should allow `this` in constructors. Because sometimes member fields and parameters have duplicate names, `this` must be used as a distinction.

**Comment by shargon** [comment]
Agree with the rule, when the name os the same, we need to use `this.`

**Comment by igormcoelho** [comment]
This logic is fine for me, using this to avoid naming conflicts. It's similar to this, but more flexible: http://stylecop.soyuz5.com/SA1101.html
There's an option to configure alternative rules too... but in this, better keep an "informal" (but explicit) core-dev rule.

File: src/neo-vm/Types/Integer.cs [hunk 1]
```diff
@@ -9,7 +9,8 @@ public class Integer : StackItem
     {
         private static readonly byte[] ZeroBytes = new byte[0];
 
-        private BigInteger value;
+        private int _length = -1;
+        private readonly BigInteger value;
```
Comment (igormcoelho) [inline src/neo-vm/Types/Integer.cs]: This is interesting. Means that Integer VM type is immutable. Agreed.

File: src/neo-vm/Types/ByteArray.cs [hunk 2]
```diff
@@ -6,7 +6,7 @@ namespace Neo.VM.Types
     [DebuggerDisplay("Type={GetType().Name}, Value={System.BitConverter.ToString(value).Replace(\"-\", string.Empty)}")]
     public class ByteArray : StackItem
     {
-        private byte[] value;
+        private readonly byte[] value;
```
Comment (igormcoelho) [inline src/neo-vm/Types/ByteArray.cs]: This is interesting. Means that ByteArray VM type is immutable. Agreed.

File: src/neo-vm/Types/Boolean.cs [hunk 3]
```diff
@@ -10,7 +10,7 @@ public class Boolean : StackItem
         private static readonly byte[] TRUE = { 1 };
         private static readonly byte[] FALSE = { 0 };
 
-        private bool value;
+        private readonly bool value;
```
Comment (igormcoelho) [inline src/neo-vm/Types/Boolean.cs]: This is interesting. Means that Boolean VM type is immutable. Agreed.

File: src/neo-vm/Types/ByteArray.cs [hunk 4]
```diff
@@ -16,7 +16,7 @@ public ByteArray(byte[] value)
         public override bool Equals(StackItem other)
         {
             if (ReferenceEquals(this, other)) return true;
-            if (ReferenceEquals(null, other)) return false;
+            if (other is null) return false;
```
Comment (igormcoelho) [inline src/neo-vm/Types/ByteArray.cs]: Is it the same?

File: src/neo-vm/Types/InteropInterface.cs [hunk 5]
```diff
@@ -17,11 +17,11 @@ public override byte[] GetByteArray()
     public class InteropInterface<T> : InteropInterface
         where T : class
     {
-        private T _object;
+        private readonly T _object;
```
Comment (igormcoelho) [inline src/neo-vm/Types/InteropInterface.cs]: This is interesting. Means that Interop VM type is immutable. Agreed.
However... is it really true? Can't upper layers somehow DUP this type, and have side effects between them?

File: src/neo-vm/Types/ByteArray.cs (hunk 4 above)
Comment (shargon) [reply src/neo-vm/Types/ByteArray.cs]: Yes, but usually we use this style

File: src/neo-vm/Types/InteropInterface.cs (hunk 5 above)
Comment (igormcoelho) [reply src/neo-vm/Types/InteropInterface.cs]: Simple example:
Syscall.GetEnumerator
Dup
Invoke something using enumerator

Question is: will enumerator be readonly for the method that needs it? If not, will some "side effect" be allowed to happen between them?
I suspect we need it non-readonly for external access.

**Pull request merged**